### PR TITLE
Tests for visitors

### PR DESCRIPTION
**Pull request by cicr99**
The propose of this set of tests is to check the correct behavior of warp visitors for ASTs. In order to do so there is the `AstTools` file, which implements:
- `AstPrinter`: prints a given ast in a readable format
- `AstParser`: takes an input in the same format mentioned above and parses it into an ast

In the `tests/ast` folder you can find three kinds of file per test:
- `<file_name>.ast` which contains the ast to be processed
- `<file_name>.ast.result` which contains the ast that is expected after the processing
- `<file_name>_test.py` which contains the script to run the test.

### CODE REVIEW

File: warp/yul/AstTools.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+        return items
+
+    def get_tabs(self):
+        tabs = 0
+        if self.pos < len(self.lines):
+            for c in self.lines[self.pos]:
+                if not c == "\t":
+                    break
+                tabs += 1
+            else:
```
[the remark this answers is not in the document]
Comment (cicr99) [reply warp/yul/AstTools.py:408]: Do you mean the `else` statement? If so, I put that `else` there for the case when you come out of the `for` naturally, in other words, when you don't hit the `break` statement. This `else` would mean that the entire line is filled with tabs, so an error is raised.

File: warp/yul/AstTools.py [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+            "Leave",
+            "Block"
+        ]
+
+        for node_type in statements:
+            parser_name = f"parse_{self.get_name(node_type)}"
+            parser = getattr(self, parser_name)
+            try:
+                return parser()
+            except AssertionError:
```
[the remark this answers is not in the document]
Comment (murcake) [reply warp/yul/AstTools.py:376]: I'm okay with merging as-is, as long as this is fixed

File: warp/yul/AstTools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (murcake) [reply warp/yul/AstTools.py:408]: Yeah

File: warp/yul/AstTools.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+
+    def parse_node(self) -> ast.Node:
+        tabs = self.get_tabs()
+        node_type_name = self.get_word(tabs).split(":")[0]
+        parser_name = f"parse_{self.get_name(node_type_name)}"
+
+        try:
+            return getattr(self, parser_name)()
+        except:
+            raise WarpException("Wrong node type name!")
```
Comment (murcake) [inline warp/yul/AstTools.py:354]: When I did this with AstVisitor, I got a lot of nasty false positives. I suggest doing the following
```suggestion
        parser = getattr(self, parser_name)
        if not parser: raise WarpException(bla-bla)
        return parser()
        
        try:
            return getattr(self, parser_name)()
        except:
            raise WarpException("Wrong node type name!")
```

**Review by JorikSchellekens — APPROVED** [review]
I agree with Artem the solidity files should go, they will get out of date with the tests and make things confusing. And the next step would be to go from yul directly. If we go from yul the solidity files could be compiled to yul in order to create the tests. Otherwise this is excellent. Can you delete the sol files and then merge?

**Pull request merged**